src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs

```diff
@@ -1,22 +1,183 @@
+    // the documentation for CFRunLoop for more information on the components involved.
     public class NetworkChange
     {
-        static public event NetworkAddressChangedEventHandler NetworkAddressChanged
+        private static object s_lockObj = new object();
+
+        // The list of current subscribers.
+        private static NetworkAddressChangedEventHandler s_addressChangedSubscribers;
+
+        // The dynamic store. We listen to changes in the IPv4 and IPv6 address keys.
+        // When those keys change, our callback below is called (OnAddressChanged).
+        private static SafeCreateHandle s_dynamicStoreRef;
+
+        // The RunLoop source, created over the above SCDynamicStore.
+        private static SafeCreateHandle s_runLoopSource;
+
+        // Listener thread that adds the RunLoopSource to its RunLoop.
+        private static Thread s_runLoopThread;
+
+        // The listener thread's CFRunLoop.
+        private static CFRunLoopRef s_runLoop;
+
+        // Use an event to try to prevent StartRaisingEvents from returning before the
+        // RunLoop actually begins. This will mitigate a race condition where the watcher
+        // thread hasn't completed initialization and stop is called before the RunLoop even starts.
+        private static readonly AutoResetEvent s_runLoopStartedEvent = new AutoResetEvent(false);
+        private static readonly AutoResetEvent s_runLoopEndedEvent = new AutoResetEvent(false);
+
+        public static event NetworkAddressChangedEventHandler NetworkAddressChanged
         {
             add
             {
-                throw new NotImplementedException();
+                lock (s_lockObj)
+                {
+                    if (s_addressChangedSubscribers == null)
+                    {
+                        CreateAndStartRunLoop();
+                    }
+
+                    s_addressChangedSubscribers += value;
+                }
             }
             remove
             {
-                throw new NotImplementedException();
+                lock (s_lockObj)
+                {
+                    bool hadSubscribers = s_addressChangedSubscribers != null;
+                    s_addressChangedSubscribers -= value;
+
+                    if (hadSubscribers && s_addressChangedSubscribers == null)
+                    {
+                        StopRunLoop();
+                    }
+                }
+            }
+        }
+
+        private static unsafe void CreateAndStartRunLoop()
+        {
+            Debug.Assert(s_dynamicStoreRef == null);
+
+            var storeContext = new Interop.SystemConfiguration.SCDynamicStoreContext();
+            using (SafeCreateHandle storeName = Interop.CoreFoundation.CFStringCreateWithCString("NetworkAddressChange.OSX"))
+            {
+                s_dynamicStoreRef = Interop.SystemConfiguration.SCDynamicStoreCreate(
+                    storeName.DangerousGetHandle(),
+                    OnAddressChanged,
```

Comment:
This doesn't look safe.  The C# compiler is going to allocate a delegate, but that'll be collected as soon as this call to create returns, at which point the native could try to call back to a collected delegate.  This delegate should be stored into a static, and then that static passed in here.